util/urlutil/redact.go

```diff
@@ -0,0 +1,33 @@
+package urlutil
+
+import (
+	"net/url"
+)
+
+const mask = "xxxxx"
+
+// RedactCredentials takes a URL and redacts username and password from it.
+// e.g. "https://user:password@host.tld/path.git" will be changed to
+// "https://xxxxx:xxxxx@host.tld/path.git".
+func RedactCredentials(s string) string {
+	ru, err := url.Parse(s)
+	if err != nil {
+		return s // string is not a URL, just return it
+	}
+	var (
+		hasUsername bool
+		hasPassword bool
+	)
+	if ru.User != nil {
+		hasUsername = len(ru.User.Username()) > 0
+		_, hasPassword = ru.User.Password()
+	}
+	if hasUsername && hasPassword {
+		ru.User = url.UserPassword(mask, mask)
+	} else if hasUsername {
+		ru.User = url.User(mask)
+	} else if hasPassword {
```

Comment:
Is this case actually valid? (password without username?)

I think in moby we take a slightly simpler approach  and just set `xxxxx:xxxxx` if any username and/or password is set (recently moved that code to be exported; first commit of https://github.com/moby/moby/pull/42835)

Reply:
> Is this case actually valid? (password without username?)

LOL, looks like I actually had a test-case for it as well in my code 😂, so I guess "yes, it's possible"